### PR TITLE
Add basic conversion between ttir and linalg

### PR DESCRIPTION
**Pull request by vwellsTT**
First step in our hoisting TTIR CPU ops into separate dynamic lib execution, split into separate PR

### CODE REVIEW

**Comment by ctodTT** [comment]
Could be wrong, but I don't see any `CMakeLists.txt`  changed here, do these new `Conversion/` files get compiled a different way?

**Comment by vwellsTT** [comment]
> Could be wrong, but I don't see any `CMakeLists.txt` changed here, do these new `Conversion/` files get compiled a different way?

Let me double-check, there were quite a few interleaved commits I tried to manually split out, so wouldn't surprise me if I missed one.  Thanks!

File: include/ttmlir/Conversion/TTIRToLinAlg/TTIRToLinAlg.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#define TTMLIR_CONVERSION_TTIRTOLINALG_TTIRTOLINALG_H
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+
+// #include "mlir/Dialect/Linalg/IR/LinalgOps.h"
```
Comment (nsmithtt) [inline include/ttmlir/Conversion/TTIRToLinAlg/TTIRToLinAlg.h:15]: Stray comment

File: lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+} // namespace
+
+namespace mlir::tt {
+
+void populateTTIRToLinAlgPatterns(MLIRContext *ctx, RewritePatternSet &patterns,
+                                  TypeConverter &typeConverter) {
+  // clang-format off
+  // ANCHOR: op_rewriter_pattern_set
+  patterns
```
[the remark this answers is not in the document]
Comment (nsmithtt) [reply lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp:93]: I think we can remove the `// clang-format off` and `// ANCHOR`, the latter is for mdbook to pick out a snippet of doc and inline it into documentation.

[the remark this answers is not in the document]
Comment (nsmithtt) [reply lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp:93]: I also think we should remove all of the commented out op conversions below.

**Review by nsmithtt — APPROVED** [review]
Minor comments inline, otherwise looks great!

File: lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      rewriter.replaceOpWithNewOp<linalg::SubOp>(
+          srcOp, adaptor.getInputs(), adaptor.getOutputs(), srcOp->getAttrs());
+
+      // Broadcast for rhs operand require the operation to be commutative to
+      // allow switching the order of operands. To allow this conversion, the
+      // following conversion is applied to SubtractOp: subtractOp(lhs,rhs) ->
+      // addOp(lhs, negOp(rhs))
+
+    } else {
+      auto negEmptyOp = rewriter.create<tensor::EmptyOp>(
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp:69]: This won't work. linalg doesn't support implicit broadcast. You should add explicit broadcast followed up with sub.

File: lib/RegisterAll.cpp [hunk 4]
```diff
@@ -36,6 +37,13 @@ void mlir::tt::registerAllDialects(mlir::DialectRegistry &registry) {
 #if TTMLIR_ENABLE_STABLEHLO
   mlir::stablehlo::registerAllDialects(registry);
 #endif
+  arith::registerBufferizableOpInterfaceExternalModels(registry);
```
Comment (mtopalovicTT) [inline lib/RegisterAll.cpp:40]: Can you comment why this is needed.

File: include/ttmlir/Conversion/TTIRToLinAlg/TTIRToLinAlg.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+
+namespace mlir::tt {
+
+void populateTTIRToLinAlgPatterns(MLIRContext *ctx, RewritePatternSet &patterns,
```
Comment (svuckovicTT) [inline include/ttmlir/Conversion/TTIRToLinAlg/TTIRToLinAlg.h:17]: I think we should use `Linalg` instead of `LinAlg`. I know it's 2 words, but if you search the llvm repo for `linalg`, there's no `LinAlg` anywhere in use in the first 2 pages of results.

File: lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    if (lhsType.getShape() == rhsType.getShape()) {
+      rewriter.replaceOpWithNewOp<linalg::SubOp>(
+          srcOp, adaptor.getInputs(), adaptor.getOutputs(), srcOp->getAttrs());
+
+      // Broadcast for rhs operand require the operation to be commutative to
+      // allow switching the order of operands. To allow this conversion, the
+      // following conversion is applied to SubtractOp: subtractOp(lhs,rhs) ->
+      // addOp(lhs, negOp(rhs))
+
+    } else {
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp:68]: Instead of the `else` block, can we `return success()` at the end of the `if` block, and then the `else` statement isn't needed.

https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html#code-denesting-inversion

File: lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        mlir::cast<RankedTensorType>(adaptor.getInputs().back().getType());
+
+    if (lhsType.getShape() == rhsType.getShape()) {
+      rewriter.replaceOpWithNewOp<linalg::SubOp>(
+          srcOp, adaptor.getInputs(), adaptor.getOutputs(), srcOp->getAttrs());
+
+      // Broadcast for rhs operand require the operation to be commutative to
+      // allow switching the order of operands. To allow this conversion, the
+      // following conversion is applied to SubtractOp: subtractOp(lhs,rhs) ->
+      // addOp(lhs, negOp(rhs))
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp:66]: Comments should always refer to code below them.

File: lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    } else {
+      auto negEmptyOp = rewriter.create<tensor::EmptyOp>(
+          srcOp.getLoc(), rhsType.getShape(), rhsType.getElementType());
+      auto negOp = rewriter.create<linalg::NegFOp>(
+          srcOp.getLoc(), ValueRange{adaptor.getInputs().back()},
+          ValueRange{negEmptyOp}, srcOp->getAttrs());
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToLinAlg/TTIRToLinAlg.cpp:73]: I think you might be able to get away without specifying `ValueRange` before initializer lists here.

File: lib/Conversion/TTIRToLinAlg/TTIRToLinAlgPass.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "ttmlir/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.h"
+#include "ttmlir/Dialect/TTIR/IR/TTIR.h"
+#include <mlir/Dialect/Func/IR/FuncOps.h>
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToLinAlg/TTIRToLinAlgPass.cpp:16]: https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html#includes

File: test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir [hunk 10]
```diff
@@ -0,0 +1,12 @@
+// RUN: ttmlir-opt --convert-ttir-to-linalg %s | FileCheck %s
+module attributes{} {
+  func.func @add(
+    %arg0: tensor<32x32xf32>,  // First input tensor
+    %arg1: tensor<32x32xf32>,  // Second input tensor
+    %arg2: tensor<32x32xf32>   // Output tensor (result stored here)
+  ) -> tensor<32x32xf32> {
+    %1 = "ttir.add"(%arg0, %arg1, %arg2) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<32x32xf32>, tensor<32x32xf32>, tensor<32x32xf32>) -> tensor<32x32xf32>
+     // CHECK: [[VAL1:%[0-9]+]] = linalg.add ins(%arg{{[0-9]+}}, %arg{{[0-9]+}} : tensor<32x32xf32>, tensor<32x32xf32>) outs(%arg{{[0-9]+}} : tensor<32x32xf32>) -> tensor<32x32xf32>
```
Comment (svuckovicTT) [inline test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir:9]: `[[VAL1:%[0-9]+]]` -> `{{%[0-9]+}}`

**Review by svuckovicTT — COMMENTED** [review]
Hey Vincent, awesome change, I'm really excited to see the CPU path working e2e!

Just a heads up, we've recently started paying more attention towards making our code as readable as possible. This guide should be useful: https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html

Also, if you have any suggestions on how we can make it even better, they would be most welcome!

I see @mtopalovicTT has covered most of the functional stuff, so I'll defer the approval to him.

File: include/ttmlir/Conversion/TTIRToLinAlg/TTIRToLinAlg.h (hunk 5 above)
Comment (vwellsTT) [reply include/ttmlir/Conversion/TTIRToLinAlg/TTIRToLinAlg.h:17]: True, I find this a little harder to parse myself, but that does seem more consistent 👍

**Comment by vwellsTT** [comment]
> Hey Vincent, awesome change, I'm really excited to see the CPU path working e2e!
> 
> Just a heads up, we've recently started paying more attention towards making our code as readable as possible. This guide should be useful: https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html
> 
> Also, if you have any suggestions on how we can make it even better, they would be most welcome!
> 
> I see @mtopalovicTT has covered most of the functional stuff, so I'll defer the approval to him.

Thanks!  One question about style guide: Milan mentioned on another PR we should prefer `llvm::` types vs `std::` types where possible--should this be added to the style guide somewhere?

File: lib/RegisterAll.cpp (hunk 4 above)
Comment (vwellsTT) [reply lib/RegisterAll.cpp:50]: Note that these shouldn't have been in this PR at all, mis-cherry-pick issue from me.  However, I just merged PR where they should be, so adding comment here anyway.

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+      }
+      broadcastedShape[rightIdx] = std::max(targetDim, inputDim);
+    }
+  }
+  return success();
+}
+
+// Helper func to check which dims need to be broadcast and which need to be
+// collapsed.  Assumes that inputShape is broadcast-able to targetShape.
+static void getDimsToBroadcastAndCollapse(
```
Comment (vwellsTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:71]: This logic is much more complicated than I originally expected, but linalg.broadcastOp asserts something like "input rank + len(dimsToBroadcastAlong) == output rank".  We want to support torch-like broadcasting, where ranks with size 1 are implicitly broadcasted => we need to do a tensor.collapseOp in such cases, which requires hairy logic.

I would appreciate careful review/re-review of this logic, especially this function 🙂

File: test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    %arg0: tensor<32xf32>,
+    %arg1: tensor<32x32xf32>,
+    %arg2: tensor<32x32xf32>
+  ) -> tensor<32x32xf32> {
+    %1 = "ttir.add"(%arg0, %arg1, %arg2) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<32xf32>, tensor<32x32xf32>, tensor<32x32xf32>) -> tensor<32x32xf32>
+    // CHECK: %{{.+}} = tensor.empty()
+    // CHECK: %{{.+}} = linalg.broadcast ins(%{{.+}} : tensor<{{.*xf32}}>) outs(%{{.+}} : tensor<{{.*xf32}}>) dimensions
+    // CHECK: %{{.+}} = linalg.add ins(%{{.+}}, %{{.+}} : tensor<{{.*xf32}}>, tensor<{{.*xf32}}>) outs(%{{.+}} : tensor<{{.*xf32}}>) -> tensor<{{.*xf32}}>
+    return %1 : tensor<32x32xf32>
+  }
```
Comment (vwellsTT) [inline test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir:63]: Simiarly, I try to test variety of broadcast + collapse cases here, but let me know if you have suggestions for other cases.  (Should I add negative cases where broadcasting fails? etc)

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+#include "llvm/Support/ErrorHandling.h"
+
+using namespace mlir;
+using namespace mlir::tt;
+
+namespace {
+
+using TensorRanks = SmallVector<int64_t, 2>;
+
+static LogicalResult computeBroadcastedShape(SmallVector<Value, 3> inputs,
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:29]: Use `OpTrait::util::getBroadcastedShape` util function to calculate broadcasted shape. It will save you all work you did here.
Also you don't need to check if operands are broadcastable since we already check this using traits for all eltwise operations using `verifyBroadcastable`.

File: test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir (hunk 12 above)
Comment (mtopalovicTT) [reply test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir:63]: Negative tests is positive here, but if you add them you won't test your code since we are already testing for broadcastable operands using traits.

**Review by mtopalovicTT — CHANGES_REQUESTED** [review]

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 13 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp]: Makes sense, I have switched to this func.  Let me know if you think I should replace this `if` check with some sort of assert etc instead given checks must have occurred at TTIR level

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+    TensorRanks &broadcastDims, SmallVector<TensorRanks, 2> &reassocIndices) {
+
+  broadcastDims.clear();
+  reassocIndices.clear();
+
+  // Identify what needs broadcasting, aligning from right
+  int targetIdx = targetShape.size() - 1;
+  int inputIdx = inputShape.size() - 1;
+
+  while (targetIdx >= 0) {
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:45]: This can be replaced with mlir::OpTrait::impl::util::getBroadcastedShape right?
Also I'm not sure this algorithm is correct, since you are appending 1's at the end of broadcastedShape instead of the beginning (I think convention is beginning).

Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:45]: I use this loop to calculate which dims of the input need to be broadcasted (and perhaps collapsed), which is not quite the same I think.  targetShape param to this func is result of `mlir::OpTrait::impl::util::getBroadcastedShape`

Comment (nsmithtt) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:45]: It's often cleaner to normalize the shapes first, to the same rank:
```python
while len(a.shape) < len(b.shape):
    a = a.insert(0, 1)                                                                                                                                                                                                                        
                                                                                 
while len(b.shape) < len(a.shape):                                        
    b = b.insert(0, 1)                                                   
     
bcast_dims = []        
for i in range(len(a.shape)):
    if a.shape[i] != b.shape[i]:                                                                                                                                                                                                              
        assert a.shape[i] == 1 or b.shape[i] == 1                                                                                                                                                                                             
        bcast_dims.append(i)
```

Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:45]: > Also I'm not sure this algorithm is correct, since you are appending 1's at the end of broadcastedShape instead of the beginning (I think convention is beginning).
I don't think that's exactly what this code is doing.  But to be honest I'm having a hard time remembering what my logic does, and I wrote this 1 month ago, so that's not a good sign :) Let me see if I can make logic clearer per Nicks suggestion, or at least add some better comments

Comment (mtopalovicTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp]: aha I see you need dim indexes not broadcasted shape, sorry for confusion

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+static void getDimsToBroadcastAndCollapse(
+    ArrayRef<int64_t> inputShape, ArrayRef<int64_t> targetShape,
+    TensorRanks &broadcastDims, SmallVector<TensorRanks, 2> &reassocIndices) {
+
+  broadcastDims.clear();
+  reassocIndices.clear();
+
+  // Create padded input shape by prepending 1s
+  SmallVector<int64_t> paddedInput;
+  const int sizeDiff = targetShape.size() - inputShape.size();
```
Comment (nsmithtt) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:43]: Can this be negative?

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 14 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp]: yeah, exactly--no worries, this is pretty confusing logic.  I found a few edge cases this still fails for I'm fixing, will update w/ more test cases in a bit once I fix them all.

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 15 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:43]: I assume the broadcast shape utility doesn't ever calculate smaller shape than input, in which case no, can't negative.  But assert sounds good to be safe

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+using namespace mlir;
+using namespace mlir::tt;
+
+namespace {
+
+using TensorRanks = SmallVector<int64_t, 2>;
+
+// Helper func to check which dims need to be broadcast and which need to be
+// collapsed.  Assumes that inputShape is broadcast-able to targetShape.
+static void getDimsToBroadcastAndCollapse(
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:34]: What do you think about splitting this function into two parts, getting broadcast dims and getting collapse vector

Something like below

```cpp
// Get the dimensions to broadcast.
//
// This function calculates the dimensions to broadcast. We assume that input
// and target shapes are broadcastable. For example if input shape is [4, 1, 3]
// and we want to broadcast to [1, 4, 5, 3], the function will return [0, 2]
// since we want to broadcast 0th and 2nd dimension of result shape.
SmallVector<int64_t> getBroadcastDims(
    ArrayRef<int64_t> inputShape, ArrayRef<int64_t> targetShape, bool &needsCollapse) {
  // Calculate the size difference
  const size_t sizeDiff = targetShape.size() - inputShape.size();

  // Create the padded input shape by prepending 1s
  SmallVector<int64_t> paddedInput(sizeDiff, 1);
  paddedInput.append(inputShape.begin(), inputShape.end());

  // Find the dimensions which needs to be broadcasted
  SmallVector<int64_t> broadcastDims;
  for (const auto& it : llvm::enumerate(llvm::zip(paddedInput, targetShape))) {
    const size_t i = it.index();
    const auto& [inputDim, targetDim] = it.value();
    // Prepanded dimensions are always broadcasted
    if (i < sizeDiff || inputDim != targetDim) {
      broadcastDims.push_back(i);
    }
    else {
      needsCollapse = true;
    }
  }

  return broadcastDims;
}

// Get the dimensions to collapse.
//
// This function calculates the dimensions to collapse. We assume that input
// and target shapes are broadcastable. linalg.broadcast requires that input
// tensor only contains dimensions that won't be broadcasted in input tensor.
// For example if input shape is [4, 1, 3] and we want to broadcast to [4, 5, 3],
// then we need to collapse the first dimension of input tensor to [4, 3].
// This function calculates the dimensions to collapse. In case above, we will
// return [[0], [1, 2]].
SmallVector<TensorRanks> getCollapseDims(ArrayRef<int64_t> inputShape,
                                     ArrayRef<int64_t> targetShape) {
    // Calculate the size difference
    const size_t sizeDiff = targetShape.size() - inputShape.size();

    // Create the padded input shape by prepending 1s
    SmallVector<int64_t> paddedInput(sizeDiff, 1);
    paddedInput.append(inputShape.begin(), inputShape.end());

    TensorRanks collapseDims;
    SmallVector<TensorRanks, 2> reassocIndexes;
    for (size_t i = sizeDiff; i < targetShape.size(); ++i) {
        const size_t inputDim = paddedInput[i];
        const size_t targetDim = targetShape[i];
        // Adjust the index to account for the prepended dimensions
        // that are not part of the input shape.
        collapseDims.push_back(i - sizeDiff);
        if (inputDim == targetDim) {
            reassocIndexes.push_back(collapseDims);
            collapseDims.clear();
        }
    }

    if (!collapseDims.empty()) {
        if (reassocIndexes.empty()) {
            reassocIndexes.push_back(collapseDims);
        } else {
            reassocIndexes.back().append(collapseDims.begin(), collapseDims.end());
        }
    }

    return reassocIndexes;
}
```

File: test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir [hunk 17]
```diff
@@ -0,0 +1,142 @@
+// RUN: ttmlir-opt --convert-ttir-to-linalg %s | FileCheck %s
+module attributes{} {
+  // CHECK: func.func @add
+  func.func @add(
+    %arg0: tensor<32x32xf32>,
+    %arg1: tensor<32x32xf32>,
+    %arg2: tensor<32x32xf32>
```
Comment (mtopalovicTT) [inline test/ttmlir/Conversion/TTIRToLinalg/ttir.mlir:7]: Can you rename this file to ttirtolinalg.mlir

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+public:
+  using OpConversionPattern<TTIROpTy>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(TTIROpTy op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+
+    // First, compute broadcasted shape from operands.
+    SmallVector<Value, 3> inputs = adaptor.getInputs();
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:127]: This is used for binary eltwise so you can use SmallVector<Value, 2> instead of SmallVector<Value, 3>

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+    SmallVector<int64_t, 4> broadcastedShape;
+    if (!OpTrait::util::getBroadcastedShape(input0Shape, input1Shape,
+                                            broadcastedShape)) {
+      return rewriter.notifyMatchFailure(
+          op, "Operands are not broadcastable--this should be impossible!");
+    }
+
+    // Replace any inputs which aren't in target shape with broadcast results
+    // which are.
+    SmallVector<Value, 4> broadcastedInputs;
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:144]: SmallVector<Value, 2>

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+    for (Value input : inputs) {
+      auto inputRankedTensorType = dyn_cast<RankedTensorType>(input.getType());
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:146]: If I'm correct here, you won't need a cast. Since you are using ttir ops they are always returning RankedTensorType,
so input.getType() should be enough.

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+      if (!inputRankedTensorType) {
+        continue;
+      }
+      Type elementType = inputRankedTensorType.getElementType();
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:150]: This is only used for EmptyOp creation, you can inline it there.

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+#include "llvm/Support/ErrorHandling.h"
+
+#include <cstdint>
+
+using namespace mlir;
+using namespace mlir::tt;
+
+namespace {
+
+using TensorRanks = SmallVector<int64_t, 2>;
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:30]: Why using alias here?

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+    ArrayRef<int64_t> input0Shape =
+        dyn_cast<RankedTensorType>(inputs[0].getType()).getShape();
+    ArrayRef<int64_t> input1Shape =
+        dyn_cast<RankedTensorType>(inputs[1].getType()).getShape();
+
+    SmallVector<int64_t, 4> broadcastedShape;
+    if (!OpTrait::util::getBroadcastedShape(input0Shape, input1Shape,
+                                            broadcastedShape)) {
+      return rewriter.notifyMatchFailure(
+          op, "Operands are not broadcastable--this should be impossible!");
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:139]: nit: fix comment

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    // Replace any inputs which aren't in target shape with broadcast results
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:142]: nit: Rewrite inputs so they are compatible by adding broadcast and optional collapse_shape op.

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 16 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:34]: I think that's a great idea, thanks.  And I like the comments you added to each func a lot too

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 20 above)
Comment (mtopalovicTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:146]: Nvm, this can be any op.

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+        if (reassocIndexes.size() != inputRankedTensorType.getShape().size()) {
+          auto collapseOp = rewriter.create<tensor::CollapseShapeOp>(
+              loc, input, reassocIndexes);
+          broadcastInput = collapseOp.getResult();
+        }
+        auto initTensor = rewriter.create<tensor::EmptyOp>(
+            loc, broadcastedShape, elementType);
+        auto broadcastOp = rewriter.create<linalg::BroadcastOp>(
+            loc, broadcastInput, initTensor.getResult(), broadCastDims);
+        for (auto result : broadcastOp.getResults()) {
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:173]: Instead of for loop you can use        
broadcastedInputs.push_back(broadcastOp.getResults().front());

**Review by mtopalovicTT — APPROVED** [review]
@vwellsTT seems good, don't wanna block this anymore.

**Comment by mtopalovicTT** [comment]
> ### Goal: The end-to-end goal is to integrate a path to compile and execute specific ops or sets of ops on the CPU.
> ### Context:
> The entire task will be split into (tentatively) 7 PRs, as follows:
> 
> 1. Hoist specific ops into isolated funcs in a separate module
> 2. **Convert TTIR ops to linalg ops within the module of hoisted funcs**
> 3. Build a pipeline to lower linalg to llvm from existing conversion passes
> 4. Translate LLVM Dialect into a dynamic library for packing into flatbuffer
> 5. Generate helper functions so that we can call all of our hoisted funcs with a common signature
> 6. Insert TTNN instructions to move operands to host before executing hoisted func, then back to device afterwards
> 7. Update ttir-to-ttnn and ttnn-to-flatbuffer pipelines to use new passes, generate dylibs, and embed them into output flatbuffers, and update update runtime to consume dylibs from flatbuffers
> 
> This PR represents the 2nd subtask above--it converts TTIR ops in the "cpu" module into their linalg equivalents. In this case, we only enable a few basic operations for now, in an attempt to get basic end-to-end conversions working before fleshing out full conversion. Definitely open to adding some other specific ops to conversion before merging, not sure which ops would make most sense.
> 
> ### Example
> Input:
> 
> ```
> module  {
>   func.func @add(%arg0: tensor<32x32xf32>, %arg1: tensor<32x32xf32>, %arg2: tensor<32x32xf32>) -> tensor<32x32xf32> {
>     %1 = "ttir.add"(%arg0, %arg1, %arg2) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<32x32xf32>, tensor<32x32xf32>, tensor<32x32xf32>) -> tensor<32x32xf32>
>     return %1 : tensor<32x32xf32>
>   }
> }
> ```
> 
> Output:
> 
> ```
> module {
>   func.func @add(%arg0: tensor<32x32xf32>, %arg1: tensor<32x32xf32>, %arg2: tensor<32x32xf32>) -> tensor<32x32xf32> {
>     %0 = linalg.add ins(%arg0, %arg1 : tensor<32x32xf32>, tensor<32x32xf32>) outs(%arg2 : tensor<32x32xf32>) -> tensor<32x32xf32>
>     return %0 : tensor<32x32xf32>
>   }
> }
> ```

@vwellsTT You don't have to add same PR header every time. Just state what changes PR adds and capture all scenarios that PR covers.
For example for this PR you could state that we will add collapse_shape and broadcast_op for eltwise ops. Why collapse_shape is needed etc.

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 18 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:127]: Yup, originally thought DPS output would count as input so I used 3, but later found I was wrong and forgot to change

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+      getDimsToBroadcastAndCollapse(inputRankedTensorType.getShape(),
+                                    broadcastedShape, broadCastDims,
+                                    reassocIndexes);
+
+      if (!broadCastDims.empty()) {
+        Value broadcastInput = input;
+        // The broadcast op requires we actually collapse any dimensions with
+        // size 1 we want to broadcast along.
+        if (reassocIndexes.size() != inputRankedTensorType.getShape().size()) {
+          auto collapseOp = rewriter.create<tensor::CollapseShapeOp>(
```
[the remark this answers is not in the document]
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:165]: 🤯

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 22 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:30]: I just find it a little tedious to update the size template arg all over the place + keep synced everywhere, so I found this convenient.  But I will remove if you don't like, no strong preference

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 23 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp:139]: I changed to "Operands are not broadcastable!", let me know if you had something else in mind

File: lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp (hunk 24 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp]: updated this, again lmk if you still don't like :)
